### PR TITLE
[PXT-765] Support for Office Formats as part of Document Type through MarkdownIT

### PR DESCRIPTION
**Pull request by pierrebrunelle**
## Add PowerPoint, Word, and Excel support via MarkItDown

**Core Integration:**
- Added `markitdown[pptx,docx,xlsx]>=0.1.3` dependency
- Extended `DocumentType.DocumentFormat` enum with `PPTX`, `DOCX`, `XLSX`
- Created `get_office_handle()` that converts office files to markdown AST via MarkItDown
- Added `_markitdown_sections()` to `DocumentSplitter` for processing converted documents
- Registered `markitdown` as optional package in `env.py`

**Design Decision:**
Office documents are converted to markdown on load and processed through the existing markdown AST pipeline. This provides consistent behavior across all document types and leverages existing splitting/metadata logic.

**Bug Fix:**
Fixed edge case in `_token_chunks()` where Unicode decode failures could cause assertion errors on certain document content.

**Testing:**
- Added 3 tests in `test_document.py` covering insert, handle creation, and splitting
- Tested with 26,971-line CalPy presentation (splits into 2,369 token chunks)

**Documentation:**
- Created cookbook recipe: `doc-extract-text-from-office-files.ipynb`
- Updated cookbook README, CONTRIBUTING.md, and STYLE_GUIDE.md

### Supported Formats
- PowerPoint: `.pptx`, `.ppt`
- Word: `.docx`, `.doc`  
- Excel: `.xlsx`, `.xls`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -71,6 +71,7 @@ dependencies = [
     "pixeltable-pgserver==0.4.0",
     "tzlocal>=5.0",
     "aiohttp>=3.9",
+    "markitdown[pptx,docx,xlsx]>=0.1.3",
```
Comment (aaron-siegel) [inline pyproject.toml]: I'd presume we do not want this is a required dependency (especially since mistune is optional and that's needed too!)

File: tests/test_document.py [hunk 2]
```diff
@@ -47,6 +47,17 @@ def valid_doc_paths(self) -> list[str]:
     def invalid_doc_paths(self) -> list[str]:
         return [get_video_files()[0], get_audio_files()[0], get_image_files()[0]]
 
+    def office_doc_paths(self) -> list[str]:
+        """Return paths to office format test documents."""
+        from pathlib import Path
+
+        docs_dir = Path(__file__).parent / 'data' / 'documents'
+        office_files = []
+        for ext in ['.pptx', '.docx', '.xlsx']:
```
Comment (aaron-siegel) [inline tests/test_document.py]: style nit: use a tuple
```
for ext in ('.pptx', '.docx', '.xslx')
```

File: tests/test_document.py [hunk 3]
```diff
@@ -47,6 +47,17 @@ def valid_doc_paths(self) -> list[str]:
...
+            for f in docs_dir.glob(f'*{ext}'):
```
Comment (aaron-siegel) [inline tests/test_document.py]: Why the loop? Why not simply
```
office_files.extend(docs_dir.glob(...))
```
Or even more concise,
```
office_files = [f for ext in (...) for f in docs_dir.glob(...)]
```

File: pixeltable/iterators/document.py [hunk 4]
```diff
@@ -472,6 +480,67 @@ def _txt_sections(self) -> Iterator[DocumentSection]:
         assert self._doc_handle.txt_doc is not None
         yield DocumentSection(text=ftfy.fix_text(self._doc_handle.txt_doc), metadata=DocumentSectionMetadata())
 
+    def _markitdown_sections(self) -> Iterator[DocumentSection]:
```
Comment (aaron-siegel) [inline pixeltable/iterators/document.py]: This looks like pure slop to me. Why does this need to be a separate method from `_markdown_sections()`? Either reuse `_markdown_sections()` and parameterize as necessary to minimize code duplication, or add a comment to explain the reasoning why this needs to be separate.

File: pixeltable/utils/documents.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import dataclasses
 import os
 
-import bs4
+import bs4  # type: ignore[import-untyped]
```
Comment (aaron-siegel) [inline pixeltable/utils/documents.py]: Why?

**Review by aaron-siegel — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (pierrebrunelle) [reply pyproject.toml]: Done. Moved to dev dependencies.

File: tests/test_document.py (hunk 2 above)
Comment (pierrebrunelle) [reply tests/test_document.py]: Done.

File: tests/test_document.py (hunk 3 above)
Comment (pierrebrunelle) [reply tests/test_document.py]: Done. Simplified to single list comprehension.

File: pixeltable/iterators/document.py (hunk 4 above)
Comment (pierrebrunelle) [reply pixeltable/iterators/document.py]: Removed. Now reuses _markdown_sections() with an optional md_ast parameter for the AST source.

File: pixeltable/utils/documents.py (hunk 5 above)
Comment (pierrebrunelle) [reply pixeltable/utils/documents.py]: Removed. Unnecessary since we have types-beautifulsoup4 installed.

File: pixeltable/iterators/document.py [hunk 6]
```diff
@@ -365,9 +373,16 @@ def process_element(el: Tag | NavigableString) -> Iterator[DocumentSection]:
         yield from process_element(self._doc_handle.bs_doc)
         yield from emit()
 
-    def _markdown_sections(self) -> Iterator[DocumentSection]:
-        """Create DocumentSections reflecting the html-specific separators"""
-        assert self._doc_handle.md_ast is not None
+    def _markdown_sections(self, md_ast: list[dict] | None = None) -> Iterator[DocumentSection]:
```
Comment (aaron-siegel) [inline pixeltable/iterators/document.py]: If we're refactoring this way (which does make sense to me), better to have md_ast be a required parameter and use a consistent pattern (passing in md_ast as an argument) across both kinds of markdown input.

File: pixeltable/iterators/document.py [hunk 7]
```diff
@@ -365,9 +373,16 @@ def process_element(el: Tag | NavigableString) -> Iterator[DocumentSection]:
...
+        """Create DocumentSections from a markdown AST.
+
+        Args:
+            md_ast: The markdown AST to process. If None, uses self._doc_handle.md_ast.
+                    This allows reuse for office formats converted to markdown via MarkItDown.
+        """
+        if md_ast is None:
```
Comment (aaron-siegel) [inline pixeltable/iterators/document.py]: With the above suggestion, these three lines of code would go away.

File: pixeltable/utils/documents.py [hunk 8]
```diff
@@ -16,6 +16,8 @@ class DocumentHandle:
     md_ast: dict | None = None
     pdf_doc: fitz.Document | None = None
     txt_doc: str | None = None
+    # For office formats converted to markdown via markitdown
+    markitdown_md_ast: dict | None = None
```
Comment (aaron-siegel) [inline pixeltable/utils/documents.py]: Alternatively, is there any reason to keep this separate rather than just using md_ast in all cases?

File: tests/test_document.py [hunk 9]
```diff
@@ -47,6 +47,13 @@ def valid_doc_paths(self) -> list[str]:
     def invalid_doc_paths(self) -> list[str]:
         return [get_video_files()[0], get_audio_files()[0], get_image_files()[0]]
 
+    def office_doc_paths(self) -> list[str]:
+        """Return paths to office format test documents."""
+        from pathlib import Path
```
Comment (aaron-siegel) [inline tests/test_document.py]: Move to top of file (there is no justification for an inline import here).

File: tests/test_document.py [hunk 10]
```diff
@@ -47,6 +47,13 @@ def valid_doc_paths(self) -> list[str]:
     def invalid_doc_paths(self) -> list[str]:
         return [get_video_files()[0], get_audio_files()[0], get_image_files()[0]]
 
+    def office_doc_paths(self) -> list[str]:
```
Comment (aaron-siegel) [inline tests/test_document.py]: Missing are test documents from all supported formats.

File: pixeltable/iterators/document.py (hunk 6 above)
Comment (pierrebrunelle) [reply pixeltable/iterators/document.py]: Done. md_ast is now a required parameter.

File: pixeltable/iterators/document.py (hunk 7 above)
Comment (pierrebrunelle) [reply pixeltable/iterators/document.py]: Done.

File: pixeltable/utils/documents.py (hunk 8 above)
Comment (pierrebrunelle) [reply pixeltable/utils/documents.py]: Done. Removed markitdown_md_ast, now using md_ast for all markdown AST cases.

File: tests/test_document.py (hunk 9 above)
Comment (pierrebrunelle) [reply tests/test_document.py]: Done.

File: tests/test_document.py (hunk 10 above)
Comment (pierrebrunelle) [reply tests/test_document.py]: Added test_presentation.pptx, test_document.docx, and test_spreadsheet.xlsx.

File: pixeltable/utils/documents.py [hunk 11]
```diff
@@ -61,14 +67,14 @@ def get_html_handle(path: str) -> bs4.BeautifulSoup:
...
-def get_markdown_handle(path: str) -> dict:
+def get_markdown_handle(path: str) -> list:
     Env.get().require_package('mistune', [3, 0])
     import mistune
 
     with open(path, encoding='utf8') as file:
         text = file.read()
     md_ast = mistune.create_markdown(renderer=None)
-    return md_ast(text)
+    return md_ast(text)  # type: ignore[return-value]
```
Comment (aaron-siegel) [inline pixeltable/utils/documents.py]: What's going on here? We changed the return type of the function, and then added a `# type: ignore` because the return type is invalid? This needs an explanation at least

File: pixeltable/utils/documents.py [hunk 12]
```diff
@@ -93,3 +99,21 @@ def get_txt(path: str) -> str:
...
+
+    # Convert office document to markdown using MarkItDown
+    md = MarkItDown(enable_plugins=False)
+    result = md.convert(path)
+    markdown_text = result.text_content
+
+    # Parse the markdown into an AST for consistent handling
+    md_parser = mistune.create_markdown(renderer=None)
+    # When renderer=None, mistune returns the AST as a list
+    return md_parser(markdown_text)  # type: ignore[return-value]
```
Comment (aaron-siegel) [inline pixeltable/utils/documents.py]: Same, why not just use the correct return type?

File: tests/test_document.py [hunk 13]
```diff
@@ -372,3 +385,93 @@ def test_doc_splitter_images(self, reset_db: None) -> None:
...
+                assert handle.format == ts.DocumentType.DocumentFormat.PPTX, path
+                assert handle.md_ast is not None, path
+            elif extension in ('.docx', '.doc'):
+                assert handle.format == ts.DocumentType.DocumentFormat.DOCX, path
+                assert handle.md_ast is not None, path
+            elif extension in ('.xlsx', '.xls'):
+                assert handle.format == ts.DocumentType.DocumentFormat.XLSX, path
+                assert handle.md_ast is not None, path
+
+    def test_office_formats_splitter(self, reset_db: None) -> None:
```
Comment (aaron-siegel) [inline tests/test_document.py]: Why is this necessary? Don't we already have a `test_document_splitter` that picks up everything in the `tests/documents` folder? Is there a particular reason why we need to special-case office documents with separate tests?

File: tests/test_document.py [hunk 14]
```diff
@@ -47,6 +48,15 @@ def valid_doc_paths(self) -> list[str]:
     def invalid_doc_paths(self) -> list[str]:
         return [get_video_files()[0], get_audio_files()[0], get_image_files()[0]]
 
+    def office_doc_paths(self) -> list[str]:
```
Comment (aaron-siegel) [inline tests/test_document.py]: Next question :) Why is this special-cased rather than just using the existing `utils.get_documents()` (which already picks up everything in tests/documents) and incorporating into the existing document tests? (Also see the below comment)

**Comment by pierrebrunelle** [comment]
Addressed all review comments:

**documents.py type annotations**: Changed `# type: ignore[return-value]` to proper `cast(list, ...)` with explanatory comments. The issue is that mistune's `create_markdown(renderer=None)` is typed to return `str | list` but actually returns `list[dict]` when renderer is None.

**Redundant office tests**: Removed the separate `office_doc_paths()` method and the three office-specific test functions (`test_office_formats_insert`, `test_office_formats_get_handle`, `test_office_formats_splitter`). The office test documents are already in `tests/data/documents/` and are picked up by `get_documents()`, so `test_doc_splitter` now covers them. Added `skip_test_if_not_installed('markitdown')` to handle environments without the package.

**Review by pierrebrunelle — COMMENTED** [review]
Done.

File: pixeltable/utils/documents.py (hunk 11 above)
Comment (pierrebrunelle) [reply pixeltable/utils/documents.py]: Done. Changed to `cast(list, ...)` with a comment explaining that mistune types it as `str | list` but actually returns `list[dict]` when `renderer=None`.

File: pixeltable/utils/documents.py (hunk 12 above)
Comment (pierrebrunelle) [reply pixeltable/utils/documents.py]: Done. Same fix - using `cast(list, ...)`.

File: tests/test_document.py (hunk 13 above)
Comment (pierrebrunelle) [reply tests/test_document.py]: Removed. The office test documents are in `tests/data/documents/` and picked up by `get_documents()`, so `test_doc_splitter` handles them. Added `skip_test_if_not_installed('markitdown')` to handle environments without the package.

File: tests/test_document.py (hunk 14 above)
Comment (pierrebrunelle) [reply tests/test_document.py]: Done. Removed `office_doc_paths()` - the office documents are picked up by `get_documents()` and tested through the existing setup.

File: pixeltable/utils/documents.py [hunk 15]
```diff
@@ -93,3 +101,19 @@ def get_txt(path: str) -> str:
...
+    Env.get().require_package('markitdown')
+    import mistune
+    from markitdown import MarkItDown
+
+    md = MarkItDown(enable_plugins=False)
+    result = md.convert(path)
+    markdown_text = result.text_content
+
+    md_parser = mistune.create_markdown(renderer=None)
+    # renderer=None returns AST as list[dict], but mistune types it as str | list
```
Comment (aaron-siegel) [inline pixeltable/utils/documents.py]: You can also do it this way (usually preferred to using `cast()`):
```
result = md_parser(markdown_text)
assert isinstance(result, list)
return result
```
and mypy will understand how to interpret it.

**Review by aaron-siegel — APPROVED** [review]

File: pixeltable/utils/documents.py (hunk 15 above)
Comment (pierrebrunelle) [reply pixeltable/utils/documents.py]: Done. Updated both functions to use `isinstance` assertion.

Comment (pierrebrunelle) [reply pixeltable/utils/documents.py]: done
